rfc/0015-multi-file-jayvee/0015-multi-file-jayvee.md

```diff
@@ -0,0 +1,116 @@
+<!--
+SPDX-FileCopyrightText: 2023 Friedrich-Alexander-Universitat Erlangen-Nurnberg
+
+SPDX-License-Identifier: AGPL-3.0-only
+-->
+
+# RFC 0015: Multi-file Jayvee
+
+| | |
+|---|---|
+| Feature Tag | `multi-file` |
+| Status | `DRAFT` | <!-- Possible values: DRAFT, DISCUSSION, ACCEPTED, REJECTED -->
+| Responsible | `georg-schwarz` | <!-- TODO: assign yourself as main driver of this RFC -->
+<!-- 
+  Status Overview:
+  - DRAFT: The RFC is not ready for a review and currently under change. Feel free to already ask for feedback on the structure and contents at this stage.
+  - DISCUSSION: The RFC is open for discussion. Usually, we open a PR to trigger discussions.
+  - ACCEPTED: The RFC was accepted. Create issues to prepare implementation of the RFC.
+  - REJECTED: The RFC was rejected. If another revision emerges, switch to status DRAFT.
+-->
+
+## Summary
+
+This RFC introduces the possibility of distributing a Jayvee program over multiple files.
+This feature will foster reuse of valuetypes, blocks, and other elements.
+Inherent to this feature is a concept of how scoping and naming is handled for nested structures.
+
+
+## Motivation
+
+Right now, Jayvee users can only pack their whole Jayvee model into one file. 
+The challenge is two-fold:
+1. Larger projects will become unmaintainable quite quickly, as the elements cannot be organized into multiple files.
+2. Without distribution to multiple files, there is no possibility to reuse models of other projects.
+
+For example, we might be enable to build libraries of valuetypes that can be reused across multiple projects instead of copying the code.
+
+## Explanation
+
+### Exporting elements
+
+For exporting elements, I propose introducing a new concept called `libraries` instead of explicitly modeling an import or export per element.
+A `library` can inhibit `Valuetype`s, `Block`s, `BlockType`s,  `Constraint`s, and `Transform`s.
+
+**Example library**
+```
+library MyDomainLibrary {
+  valuetype MyDomainSpecificValuetype {
+    // ... details of valuetype
+  }
+
+  // ... possibly more elements
+}
+```
+
+Libraries are always "exported" and can be "imported" into other files.
+
+
+### Visibility of elements in a file
```

Comment:
I think we could consider, similarly to java, to force a match between files and language concepts. E.g. a .jv file can only contain one library or a pipeline, not both, not two libraries. That would create nice, clear 'library files' that only contain one library, nothing else and 'pipeline files' that contain a pipeline and optionally one of definitions of blocks and valuetypes. I'd consider that structure worth enforcing.

Reply:
While I generally agree that enforced structure might help users, I am reluctant to agree here with the Java example. It would lead to very bloated code, and in combination with the suggestion to remove block definitions from a `pipeline`, it would require jumping a lot of back and forth if you want to write reusable code.